From 309a42991e2279fdf4911f3c9084ddb84eeaf7ad Mon Sep 17 00:00:00 2001
From: Philippe Gerum <rpm@xenomai.org>
Date: Sun, 4 Jul 2021 17:24:30 +0200
Subject: [PATCH 0740/1010] evl/wait: add memory shortage as a wait abort cause

In some circumstances, a wait operation cannot complete because we
don't have enough memory to prepare the waiter for wake up, such as
transferring a buffer to it.

In this case, we want evl_wait_schedule() to notify the caller
explicitly by a distinct wake up status (i.e. -ENOMEM), which cannot
be misinterpreted as a different wake up cause.

Signed-off-by: Philippe Gerum <rpm@xenomai.org>
---
 include/evl/thread.h      |  2 +-
 include/uapi/evl/thread.h |  1 +
 kernel/evl/wait.c         | 11 ++++++++---
 3 files changed, 10 insertions(+), 4 deletions(-)

diff --git a/include/evl/thread.h b/include/evl/thread.h
index 41a7f512f5..ad198b1d20 100644
--- a/include/evl/thread.h
+++ b/include/evl/thread.h
@@ -29,7 +29,7 @@
 /* All bits which may cause an EVL thread to block in oob context. */
 #define EVL_THREAD_BLOCK_BITS   (T_SUSP|T_PEND|T_DELAY|T_WAIT|T_DORMANT|T_INBAND|T_HALT|T_PTSYNC)
 /* Information bits an EVL thread may receive from a blocking op. */
-#define EVL_THREAD_INFO_MASK	(T_RMID|T_TIMEO|T_BREAK|T_WAKEN|T_ROBBED|T_KICKED|T_BCAST)
+#define EVL_THREAD_INFO_MASK	(T_RMID|T_TIMEO|T_BREAK|T_WAKEN|T_ROBBED|T_KICKED|T_BCAST|T_NOMEM)
 /* Mode bits configurable via EVL_THRIOC_SET/CLEAR_MODE. */
 #define EVL_THREAD_MODE_BITS	(T_WOSS|T_WOLI|T_WOSX|T_HMSIG|T_HMOBS)
 
diff --git a/include/uapi/evl/thread.h b/include/uapi/evl/thread.h
index 6152904652..9f23455be6 100644
--- a/include/uapi/evl/thread.h
+++ b/include/uapi/evl/thread.h
@@ -55,6 +55,7 @@
 #define T_PTSIG   0x00001000 /* Ptrace signal is pending */
 #define T_PTSTOP  0x00002000 /* Ptrace stop is ongoing */
 #define T_PTJOIN  0x00004000 /* Ptracee should join ptsync barrier */
+#define T_NOMEM   0x00008000 /* No memory to complete the operation */
 
 /* Local information flags (private to current thread) */
 
diff --git a/kernel/evl/wait.c b/kernel/evl/wait.c
index 715b82a0ac..83144c9e06 100644
--- a/kernel/evl/wait.c
+++ b/kernel/evl/wait.c
@@ -164,9 +164,11 @@ int evl_wait_schedule(struct evl_wait_queue *wq)
 	 * stale: -EIDRM is returned.
 	 *
 	 * - if neither T_TIMEO or T_BREAK are set, we got a wakeup
-	 * and success is returned (zero). In addition, the caller may
-	 * need to check for T_BCAST if the signal is not paired with
-	 * a condition but works as a pulse instead.
+	 * and success (zero) or -ENOMEM is returned, depending on
+	 * whether T_NOMEM is set (i.e. the operation was aborted due
+	 * to a memory shortage). In addition, the caller may need to
+	 * check for T_BCAST if the signal is not paired with a
+	 * condition but works as a pulse instead.
 	 *
 	 * - otherwise, if any of T_TIMEO or T_BREAK is set:
 	 *
@@ -188,6 +190,9 @@ int evl_wait_schedule(struct evl_wait_queue *wq)
 	if (info & T_RMID)
 		return -EIDRM;
 
+	if (info & T_NOMEM)
+		return -ENOMEM;
+
 	if (info & (T_TIMEO|T_BREAK)) {
 		raw_spin_lock_irqsave(&wq->lock, flags);
 		if (!list_empty(&curr->wait_next)) {
-- 
2.30.2

